_posts/2022-01-20-making-github-action.md

```diff
@@ -0,0 +1,131 @@
+
+Ideally, then, our repositories would contain both xlsx files and their exported csv counterparts, so that both spreadsheet users and programmers could easily access their preferred format. However, asking users to manually update any changes to both formats before committing their files is both error-prone and possibly hard to maintain.
+
+GitHub Actions are one possible solution to maintaining this consistency. While GitHub Actions are often used for maintaining development workflows, including integration, deployment, and testing, there's no reason that we can't use Actions to automatically trigger any other required scriptable activity. For example, we can use an Action to maintain the connection between a repository's xlsx files and their csv counterparts.
+
+I created an action to do just that: https://github.com/CanDIG/xlsx2csv-action. Fundamentally, an Action is an executable, self-contained container that runs some code. The container will be executed by GitHub in an enclosing container, allowing each action's container to access the enclosing environment.The Action uses a manifest file to describe how its container will interact with the hosting container. With a GitHub Action, we can guarantee that we are executing in an environment with all of the required packages installed. Environment variables are defined to access inputs and the common workspace.
+
+So, to begin creating this action, we write a simple Python script that performs the conversion on an inputted file. This Python file, at its most basic, uses the xlsx2csv package to convert xlsx files to csvs. We can do a little more than that: the script can walk through the repository's files (temporarily stored in the action's workspace) and convert all xlsx files it finds.
+
+```
+        dirlist = [args.input]
+        files = []
+        while len(dirlist) > 0:
+            for (dirpath, dirnames, filenames) in os.walk(dirlist.pop()):
+                dirlist.extend(dirnames)
+                files.extend(map(lambda n: os.path.join(*n), zip([dirpath] * len(filenames), filenames)))
+        for file in files:
+            file_match = re.match("(.+)\.xlsx$", file)
+            if file_match is not None:
+                output_dir = f"{os.environ.get('GITHUB_WORKSPACE')}/{file_match.group(1)}"
+                Xlsx2csv(file, outputencoding="utf-8").convert(output_dir, sheetid=0)
+```
+
+We can then wrap this in the standard shell script for GitHub Actions, entrypoint.sh:
+
+```
+#!/bin/sh
+
+python /convert.py $*
+```
+
+And then execute that script from a Dockerfile:
+
+```
+FROM python:3.9
+
+COPY . /
+
+RUN pip install -r requirements.txt
+
+ENTRYPOINT ["/entrypoint.sh"]
+```
+
+The nice thing about using Docker is that you can easily create a Docker environment on your local machine that replicates the environment that the Action will execute in. It'll look something like this docker-compose.yml:
+
+```
```

Comment:
I think it's \`\`\`yaml although maybe there's a specific docker-compose highlight